src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java

```diff
@@ -0,0 +1,273 @@
+                    String errorMessage = format("Error sharing Instance with id %s to the target tenant %s. Because source is %s",
+                      instanceId, sharingInstance.getTargetTenantId(), srcInstance.getSource());
+                    sendErrorResponseAndPrintLogMessage(tenantId, errorMessage, sharingInstance, record.headers());
+                    promise.fail(errorMessage);
+                  } else {
+                    String errorMessage = format("Error sharing Instance with id %s to the target tenant %s. Because source is %s",
+                      instanceId, sharingInstance.getTargetTenantId(), srcInstance.getSource());
+                    sendErrorResponseAndPrintLogMessage(tenantId, errorMessage, sharingInstance, record.headers());
+                    promise.fail(errorMessage);
+                  }
+                }
+              })
+              .onFailure(err -> {
+                String errorMessage = format("Error retrieving Instance by id %s from source tenant %s. Error: %s", instanceId, sharingInstance.getSourceTenantId(), err);
+                sendErrorResponseAndPrintLogMessage(tenantId, errorMessage, sharingInstance, record.headers());
+                promise.fail(errorMessage);
+              });
+          } else {
+            String errorMessage = format("Error checking Instance by id %s on target tenant %s. Error: %s", instanceId, sharingInstance.getTargetTenantId(), failure.getMessage());
+            sendErrorResponseAndPrintLogMessage(tenantId, errorMessage, sharingInstance, record.headers());
+            promise.fail(errorMessage);
+          }
+        })
+        .onSuccess(instanceOnTargetTenant -> {
+          String errorMessage = format("Instance %s is present on target tenant: %s", instanceId, sharingInstance.getTargetTenantId());
+          sendErrorResponseAndPrintLogMessage(tenantId, errorMessage, sharingInstance, record.headers());
+          promise.fail(errorMessage);
+        });
+      return promise.future();
+    } catch (Exception ex) {
+      LOGGER.error(format("Failed to process data import kafka record from topic %s", record.topic()), ex);
+      return Future.failedFuture(ex);
+    }
+  }
+
+  private void sendErrorResponseAndPrintLogMessage(String tenantId, String errorMessage, SharingInstance sharingInstance, List<KafkaHeader> kafkaHeaders) {
+    LOGGER.error("handle:: {}", errorMessage);
+    sendEventToKafka(tenantId, sharingInstance, ERROR, errorMessage, kafkaConfig, kafkaHeaders);
+  }
+
+  private Future<Instance> publishInstanceToTenant(Instance instance, InstanceCollection instanceCollection) {
+    JsonObject jsonInstance = instance.getJsonForStorage();
+    jsonInstance.remove("hrid");
+    return addInstance(Instance.fromJson(jsonInstance), instanceCollection);
+  }
+
+  private Future<Instance> getInstanceById(String instanceId, String tenantId, InstanceCollection instanceCollection) {
```

Comment:
Thinking the logic within this method should already be included in `instanceCollection.findById` and there would be no need to declare this method.